src/main/scala/strawman/collection/immutable/ImmutableArray.scala

```diff
@@ -0,0 +1,102 @@
+import strawman.collection.{IterableFactory, IterableOnce, Iterator, View}
+
+import scala.{Any, Boolean, Int, Nothing}
+import scala.runtime.ScalaRunTime
+import scala.Predef.{???, intWrapper}
+
+/**
+  * An immutable array.
+  *
+  * Supports efficient indexed access and has a small memory footprint.
+  */
+class ImmutableArray[+A] private (private val elements: scala.Array[Any]) extends IndexedSeq[A] with SeqOps[A, ImmutableArray, ImmutableArray[A]] {
+
+  protected[this] def fromIterable[E](it: strawman.collection.Iterable[E]): ImmutableArray[E] = ImmutableArray.fromIterable(it)
+
+  protected[this] def fromSpecificIterable(coll: strawman.collection.Iterable[A]): ImmutableArray[A] = ImmutableArray.fromIterable(coll)
+
+  def length: Int = elements.length
+
+  override def knownSize: Int = elements.length
+
+  def apply(i: Int): A = ScalaRunTime.array_apply(elements, i).asInstanceOf[A]
+
+  def iterator(): Iterator[A] = view.iterator()
+
+  override def map[B](f: A => B): ImmutableArray[B] = ImmutableArray.tabulate(length)(i => f(apply(i)))
+
+  override def flatMap[B](f: A => IterableOnce[B]): ImmutableArray[B] = ImmutableArray.fromIterable(View.FlatMap(coll, f))
+
+  def :+ [B >: A](elem: B): ImmutableArray[B] = {
+    val dest = scala.Array.ofDim[Any](length + 1)
+    java.lang.System.arraycopy(elements, 0, dest, 0, length)
+    dest(length) = elem.asInstanceOf[Any]
+    new ImmutableArray(dest)
+  }
+
+  override def concat[B >: A](xs: IterableOnce[B]): ImmutableArray[B] =
+    xs match {
+      case bs: ImmutableArray[B] =>
+        val dest = scala.Array.ofDim[Any](length + bs.length)
+        java.lang.System.arraycopy(elements, 0, dest, 0, length)
+        java.lang.System.arraycopy(bs.elements, 0, dest, length, bs.length)
+        new ImmutableArray(dest)
+      case _ =>
+        ImmutableArray.fromIterable(View.Concat(coll, xs))
+    }
+
+  override def zip[B](xs: IterableOnce[B]): ImmutableArray[(A, B)] =
+    xs match {
+      case bs: ImmutableArray[B] =>
+        ImmutableArray.tabulate(length min bs.length) { i =>
+          (apply(i), bs(i))
+        }
+      case _ =>
+        ImmutableArray.fromIterable(View.Zip(coll, xs))
+    }
+
+  override def filter(p: A => Boolean): ImmutableArray[A] = ImmutableArray.fromIterable(View.Filter(coll, p))
+
+  override def partition(p: A => Boolean): (ImmutableArray[A], ImmutableArray[A]) = {
+    val pn = View.Partition(coll, p)
+    (ImmutableArray.fromIterable(pn.left), ImmutableArray.fromIterable(pn.right))
+  }
+
+  override def take(n: Int): ImmutableArray[A] = ImmutableArray.tabulate(n)(apply)
+
+  override def drop(n: Int): ImmutableArray[A] = ImmutableArray.tabulate((length - n) max 0)(i => apply(n + i))
+
+  override def tail: ImmutableArray[A] =
+    if (length > 0) {
+      val dest = scala.Array.ofDim[Any](length - 1)
+      java.lang.System.arraycopy(elements, 1, dest, 0, length - 1)
+      new ImmutableArray(dest)
+    } else ???
```

Comment:
That goes for other cases, too. I left it at `???` for now because we're doing the same in `List`. We need to check for all occurrences and replace them with the correct exception.